### DISCUSSION TITLE
route all traffic to same webserver without connecting to real

### DISCUSSION THREAD

**Discussion by azertyalex** [mitmproxy/mitmproxy#5416] [Q&A] [answered]
I need to use mitmproxy in a rather weird setup.

First of all, all clients need to use mitmproxy as a http/socks proxy, where certificates are generated by mitm.
This is no difficult task since it can do this already and we have access to all clients. We can import CA and set the browser settings.

Second, all incoming requests (be it _example.com_, _google.com_, ...) must be routed to the same webserver with a fixed domain name. This again is no difficult task since mitm can do this already with --mode upstream

However, I cant seem to get both scenario's working at the same time.

What I tried so far:

When using mitmdump -s main.py, it always creates a socket connection to the original servers domain name (example.com). I don't want this. Sometimes it's not even possible because a requested domain would just not exist. Which causes all to fail.
After further reading I found that I could use the upstream
command used:
```bash
mitmdump -s main.py --mode upstream:"mydomain.com" --set proxy_debug -vvv
```
code used:
```python
from mitmproxy import http, websocket, ctx, proxy
import requests
import os

class Proxy:
    def server_connect(self, data: proxy.server_hooks.ServerConnectionHookData):
        print("server_connect", data.server.address)
        print(data.server.sni)
        data.server.sni = "mydomain.com"

    def request(self, flow: http.HTTPFlow) -> None:
        flow.request.headers["x-mitm"] = "true"
        flow.request.host = "mydomain.com"
        if (flow.request.scheme == "http"):
            flow.response = http.Response.make(
                301,
                "",
                {"Location": flow.request.pretty_url.replace("http", "https")}
            )

addons = [
    Proxy()
]
```

But the error `Upstream proxy mydomain.com:443 refused HTTP CONNECT request: 400 Bad Request`. Which seems correct, because it is no proxy but a webserver/reverse proxy.

I'm not really sure what to try anymore

**Comment by azertyalex** [comment, marked as answer]
We managed to fix it by adding `--set upstream_cert=false`

**Reply by mhils** [reply]
Thanks for posting this here! Your feedback has been helpful, I've just opened a PR that makes sure that setting `connection_strategy` to `lazy` is enough and `upstream_cert=false` is not required anymore (https://github.com/mitmproxy/mitmproxy/pull/5487).

**Comment by xmo-odoo** [comment]
Complement in case somebody else with a variation of this issue arrives here: i've had the same issue using `--map-remote` and a transparent proxy (to transparently reroute connections from a real remote to a mock remote).

This behaviour has been plaguing me as the real remote is kinda flaky (one of the reasons to mock it), so mitmproxy's upstream certificate sniffing would lead to unnecessary connection failures:
```
127.0.0.1:49008: error establishing server connection: [Errno 110] Connect call failed ('...', 443)
127.0.0.1:49008: CONNECT api.remote.com:443 HTTP/1.0
     << HTTP/1.1 502 Bad Gateway 91b
127.0.0.1:49008: client disconnect
```
as well as hammering the real remote.

The solution provided here (`upstream_cert: false`) *was not sufficient*: mitmproxy would still established connections every time. However it gave me the tools to dig further and find #5123 and the second missing piece: `connection_strategy: lazy`.

With that, mitmproxy has stopped contacting the upstream entirely, increasing reliability and throughput.

**Reply by azertyalex** [reply]
Connection_strategy was already something we had found and are using. I don't really know why I didn't mention it in my original post.
